### ISSUE TITLE
Reproduce the results of the paper: Collaborative Deep learning for Recommender Systems

### ISSUE THREAD

**Issue by xanhho** [js05212/CDL#4]
Hi @js05212,

I tried to reproduce the results in your paper, but the results are not expected.
For citeulike-a dataset, sparse setting, I get the results like this:
M = 50, 100, 150, 200, 250, 300
50: 0.114028
100: 0.166543
150: 0.204587
200: 0.233036
250: 0.256955
300: 0.276866 
I compare this result with Figure-4 in your paper, it worse than.
Can you describe how you get the results in the paper? After training process, what steps you do?

**Comment by js05212** [comment]
Hi,

Did you turn on the pretrain switch and what hyperparmeters did you use?

Best,
Hao

**Comment by xanhho** [comment]
Hi,
I try to run file example/cdl_main.m, when I run with para_pretrain = 0, there is no problem. 
The results I had posted is the results get from run cdl_main.m with para_pretrain = 0.

When I try to run with your default code, para_pretrain = 1, I got the problem "Error load: unable to find file cdl1" in the line "load(my.pretrain)" of the file cdl_worker.m
if my.pretrain
    load(my.pretrain);
    fprintf('using pretrain result: %s\n',my.pretrain);
end
Can you guide me how to fix this problem?
Thank you very much for your support!

**Comment by js05212** [comment]
Oh, by pretrain switch I mean 'do_pretrain = 1;' in the cdl_worker.m file, which should be on by default.

And what hyperparameters did you use?

Hao

**Comment by xanhho** [comment]
I use the hyperparameters the same with hyperparameters in your file (cdl_worker.m and cdl_main.m)
my.folder = fd; # fd = 45 
my.weight_decay = 1e-4; % default is 1e-4
my.dropout = dpt; # 0.1
my.iter = 1;
do_pretrain = 1;
my.save_lag = sl;
my.sdae.n_epoch = nepo;
my.dae.n_epoch = 500; #500
my.max_iter = 100;


para_gpuDevice=4;
para_layers=[8000 200 50];
para_blayers=[1 1 1];
para_lv=10;
para_lu=1;
para_ln=1e3;
para_pretrain=0; #1 
para_save=999;
para_folder=45;
para_dropout=0.1;
para_from=1;
para_sdae_n_epoch=1000;#1000
para_save_lag=501;

I just change the only parameter para_pretrain in  cdl_main.m to fix the problem.

**Comment by xanhho** [comment]
Can you please tell me what should I do to reproduce the results in your paper?

**Comment by js05212** [comment]
Hi,

What version of GSL are you using? And how do you calculate the recall? There might be some differences here.

Best,
Hao

**Comment by xanhho** [comment]
Hi,
I am using gsl-2.2. I calculate the recall as the following steps: 
- R_true = read_user('ctr-data/folder45/cf-test-1-users.dat') 
- R = U*V.T
- For each user:
   + Sort list rating in R, cut at M items (M = 50, 100, ..., 300)
   + Recall for user i = (R_true[i] intersection R_cut[i] / len(R_true[i])
- The final recall is average of all users

- I also have a question: 
For the dataset CiteUlike-a, when we use the training set with P = 10 (dense setting).
In the file "cf-train-1-users.dat", we have some users with no relevant items (the number of relevant items = 0, about 412 users).
For this case, when calculating the Map or Recall score, we have to divide by the number of users. What is the number of users that the paper chooses? (num_u = 5551 or num_u = 5551 - 412)

How about you? How do you calculate the recall?
Thank you very much for your support!

Best

**Comment by js05212** [comment]
Hi, sorry for the delayed reply! I think the process is correct. And to calculate the average recall, I think we should use num_u = 5551 - 412. Is the gap because of this?

**Comment by xanhho** [comment]
Hi, thank you for your reply! 
Here is an example of the gap when I using MXNet-for-CDL:
  	     5139	        5551	
50 : 0.196949	   0.182331 
100: 0.287986	    0.266611
150: 0.352273	    0.326127
200: 0.400290	    0.370580
250: 0.438390	    0.405852 
300: 0.467425	    0.432733
For CDL, I also not yet get the results like the results in your paper.
Have you used file training data for evaluating? (There are some papers cited your paper and the authors of some papers said that they used training data for evaluating, https://github.com/eelxpeng/CollaborativeVAE/issues/2#issuecomment-337910426)

**Comment by js05212** [comment]
It looks weird to use training data for evaluation. In our case, the recall is calculated based on the test dat file, the training dat file is used only to exclude the training items from the cadidate sets. The numbers in the CVAE issue page does look much higher than CDL.

Best,
Hao

**Comment by js05212** [comment]
For the MXNet CDL, you might need to add dropout and denoising and pretraining to match the results in CDL. For CDL, what are the numbers when you use 5139?

**Comment by xanhho** [comment]
I also feel weird if use the training data for evaluation, so I try to change something and ask you. But, for
- CDL (P=1), the results I get it's not expected ... I running on Octave, I think it will fine, but I don't know why I cannot get the results as in the paper.
- CDL (P=10): I have some errors related to the Octave systems now (not related to your code), so I'm not finish training process. I will try to calculate the recall with the number of users 5139.

**Comment by js05212** [comment]
Hi,

Could you tell me what the recall is for P=1 on Octave? Is it close or is it very far away?

Best,
Hao

**Comment by xanhho** [comment]
Hi,

In the first question of this threat, I had shown the results, this is the results I get on Octave.
For citeulike-a dataset, sparse setting (P = 1), I get the results like this:
M = 50, 100, 150, 200, 250, 300
50: 0.114028
100: 0.166543
150: 0.204587
200: 0.233036
250: 0.256955
300: 0.276866
I compare this result with Figure-4 in your paper, it worse than.

**Comment by js05212** [comment]
Oh I see. I thought 0.276866 (shown at the start of this issue) is the recall@300 with the number of users **5551** (if so the correct recall would be close to 30).

**Comment by xanhho** [comment]
Yes, for the case sparse setting (P = 1), the number of users is 5551. 
I compare the results with Figure-4, I see the results I get is worse than the results in your paper.

**Comment by js05212** [comment]
Hi,

Sorry for not being clear. I mean if you divide the total recall with (5551-412), you would get an average recall of about 0.30 rather than 0.27. Is that correct?

Hao

**Comment by xanhho** [comment]
In my view:
- For the case sparse setting (P = 1), in the dataset (testing set), we don't have users without items, so we can not get the number of users (5551 - 412). The number of users for P = 1 setting is 5551.

- For the case dense setting (P = 10), the number of users is (5555 - 412), because, in the testing set, we have 412 users without items.

Therefore, the number of users for the case sparse setting (P = 1) is 5551, dense setting (P = 10) is 5551 - 412

**Comment by js05212** [comment]
Oh I think you are right. So there is a gap in the recall. I guess it might be due to (1) the different implementations in different versions of GSL (2) the difference between matlab and Octave. Maybe you can try GSL v1.16 (this is the version of GSL I used if I remember correctly) and try different random seeds.

Best,
Hao

**Comment by xanhho** [comment]
Thank you for your advice, I will try to change the version of GSL and try different random seeds.

**Comment by supermanvista** [comment]
@xanhxanh94 Did you get the same results in dense setting(P=10) as the original papers using the original MATLAB code ?
Did you try the other implementations using tensorflow https://github.com/gtshs2/Collaborative_Deep_Learning ? Why did I get rather low results(recall) than the paper presents? I used the exacted same hyperparameter listed in the original paper.
I seems to have a hard time to replicate the results in the original paper.
